### ISSUE TITLE
Feature request: non-directory based hierarchy and example post-processing

### ISSUE THREAD

**Issue by pelson** [sphinx-gallery/sphinx-gallery#354]
Thanks for this package! Having been helped massively by @choldgraf to get cartopy using sphinx-gallery, we have been able to delete complex sphinx code and replace it with a very simple configuration update. Awesome! 🎉 

http://scitools.org.uk/cartopy/docs/latest/gallery/

In order to group similar examples, cartopy took the approach of using metadata in the example file itself, rather than using the directory structure. This has two advantages for cartopy:

 * it is easy to change the group in which the example lives without breaking the link (because the files remains in one place, it is just listed under a different section)
 * it is possible to put the example under multiple sections at the same time

The ability to avoid *moving* the examples becomes particularly important with cartopy, as we have taken the approach of making the examples importable (and thus a pseudo part of the API).

In order to maintain these capabilities I therefore extended sphinx-gallery to pre-process the cartopy examples in order to identify subsections by looking at the magic ```__tags__ = ['abc']``` line in each of the examples.

Next, I needed to modify the rendering in order to remove the ```__tags``` line from the final output.

Achieving this cleanly was quite a challenge, and required me to Copy&Paste a few of the key functions (from v0.1.12) and monkey-patch those in to sphinx-gallery at sphinx app setup time.

The result can be seen at https://github.com/SciTools/cartopy/blob/master/docs/source/sphinxext/pre_sphinx_gallery.py.

Long-term, I'd love to be able to make these capabilities generic and a core part of sphinx-gallery. Does that align with the sphinx-gallery project goals? If so, I'd be happy to help with the implementation if that is useful.

**Comment by lesteve** [comment]
Great to hear that you found sphinx-gallery useful and that you managed to get rid of complex sphinx code, that was probably one of the goal of starting sphinx-gallery in the first place!

It seems to me that what some of your use cases are slightly off the beaten path, maybe having an example in different subsections is the one that could be needed more broadly. Reorganising the subsections without breaking the links is nice as well but I feel like this does not happen super often although it is probably less true for less mature projects. 

If the generation of subsections and the grouping of examples under these subsections could be done in a user specified function that would seem like one way of tackling most of your use cases. An alternative way of looking at it would be: "how could we make sphinx-gallery more modular so that it is easier to reuse parts of sphinx-gallery to build what you want ?".

About removing the `__tags__` line I guess it could be done although I don't see it as something extremely harmful to leave it in the examples.

I am very happy to hear what others think!

**Comment by choldgraf** [comment]
> maybe having an example in different subsections is the one that could be needed more broadly.

+1

> how could we make sphinx-gallery more modular so that it is easier to reuse parts of sphinx-gallery to build what you want

+ 100

:-)

**Comment by larsoner** [comment]
I agree making things more modular (image capturing, etc.) is the way to go. So far we have found ways of doing this that maintains the status quo but allows for more flexibility. If subsection organization and so forth can be split out into a user-override-able function, with existing functionality being the default one, that sounds like a clean way to go.

**Comment by larsoner** [comment]
(Even in cases where these things are going to be used by only one or two projects, which may or may not be the case here, modularization could and should simultaneously lead to cleaner code, which is a nice side benefit. I saw this a bit while working toward separating image scraping.)
